<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>BiocPyInterop • BiocPyInterop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="BiocPyInterop">
<meta property="og:description" content="This package coordinates use of BiocHail and
    scviR for workshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">BiocPyInterop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/BiocPyInterop.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/vjcitn/BiocPyInterop" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="biocpyinterop">BiocPyInterop<a class="anchor" aria-label="anchor" href="#biocpyinterop"></a>
</h1></div>
<p>This package is the basis of a workshop that will address use of python software in Bioconductor. We focus on use of basilisk in packages BiocHail and scviR.</p>
<div class="section level3">
<h3 id="description-of-workshop">Description of workshop<a class="anchor" aria-label="anchor" href="#description-of-workshop"></a>
</h3>
<p>Abstract: Multilingual data science strategies can increase efficiency of discovery by taking advantage of diverse data management and analysis strategies.</p>
<p>In this workshop we will examine interplay between R, Python, and Apache Spark in genetic and single-cell applications. CITE-seq studies simultaneously quantify surface protein and mRNA abundance in single cells. We will use scviR to compare interpretations based on deep learning and sequential component-specific methods.</p>
<p>The UK Biobank is the foundation of thousands of genome-wide association studies. The Telomere-to-Telomere project produced the first gapless human reference genome. Both of these resources will be explored using BiocHail. Workshop attendees will acquire an understanding of Aaron Lun’s <a href="https://bioconductor.org/packages/basilisk" class="external-link">basilisk</a> package and its use in isolating specific collections of python modules, the anndata representations and scvi-tools analyses of CITE-seq data, and the hail.is approach to structuring and analyzing massive genetics data resources using Spark Resilient Distributed Data. All programming will be carried out in R; quarto documents that mix R and python will also be illustrated.</p>
</div>
<div class="section level3">
<h3 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h3>
<ul>
<li>Basic knowledge of R syntax</li>
<li>Interest in single-cell genomics, human genetics, deep learning</li>
</ul>
<p>It will be helpful to have an acquaintance with</p>
<ul>
<li><a href="http://bioconductor.org/books/3.17/OSCA.advanced/integrating-with-protein-abundance.html" class="external-link">a chapter of the OSCA book</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/scviR/inst/doc/citeseq_tut.html" class="external-link">an scviR vignette</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/BiocHail/inst/doc/gwas_tut.html" class="external-link">a BiocHail vignette</a></li>
<li>
<a href="https://vjcitn.github.io/BiocT2T/" class="external-link">a look at BiocT2T</a> – note that we will not work with the T2T 1KG extract in detail, as it involves a 40+ GB download, but the mechanics of working with it on your own will be explained</li>
</ul>
</div>
<div class="section level3">
<h3 id="participation">Participation<a class="anchor" aria-label="anchor" href="#participation"></a>
</h3>
<p>This is a 90 minute workshop that will cover</p>
<ul>
<li>programming with basilisk to establish predictable python infrastructure and interoperation</li>
<li>exploration of torch-based tooling for single-cell analysis of a CITE-seq experiment</li>
<li>exploration of spark-based tooling for interaction with 1000 genomes genotypes (and, if time permits, UK Biobank phenotypes)</li>
</ul>
</div>
<div class="section level3">
<h3 id="r--bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r--bioconductor-packages-used"></a>
</h3>
<ul>
<li>basilisk</li>
<li>OSCA.advanced</li>
<li>scviR</li>
<li>BiocHail</li>
</ul>
</div>
<div class="section level3">
<h3 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h3>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Verify setup</td>
<td>5m</td>
</tr>
<tr class="even">
<td>Motivations/discussion</td>
<td>5m</td>
</tr>
<tr class="odd">
<td>basilisk and basilisk.utils</td>
<td>10m</td>
</tr>
<tr class="even">
<td>(concerns with bloat)</td>
<td></td>
</tr>
<tr class="odd">
<td>OSCA advanced: CITE-seq</td>
<td>15m</td>
</tr>
<tr class="even">
<td>break</td>
<td>10 m</td>
</tr>
<tr class="odd">
<td>scviR – AnnData and tutorial VAE</td>
<td>20m</td>
</tr>
<tr class="even">
<td>Hail: exploring 1KG and UKBB</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Review</td>
<td>5m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h3>
<ul>
<li>Learning goals:<ul>
<li>understand basic issues with connecting R/Bioconductor to Python software tools for genomics</li>
<li>relate aspects of the anndata class to aspects of SummarizedExperiment</li>
<li>compare findings in a stepwise analysis of PBMC data in Bioconductor to those obtained with findings from fitting an autoencoder to the same data in scvi-tools</li>
<li>explore Hail’s structures and methods for working with genotypes and phenotypes at scale</li>
</ul>
</li>
<li>Learning objectives:<ul>
<li>Understand the user situation when basilisk manages python resources and usage<ul>
<li>review functions in basilisk.utils</li>
<li>assess resource consumption (disk space used per version of basilisk/client package)</li>
<li>process management: check ?getBasiliskFork</li>
</ul>
</li>
<li>Review an analysis of CITE-seq data on 6800 cells with Bioconductor in <a href="http://bioconductor.org/books/3.17/OSCA.advanced/integrating-with-protein-abundance.html" class="external-link">OSCA advanced</a><ul>
<li>review <a href="http://bioconductor.org/books/3.17/OSCA.advanced/integrating-with-protein-abundance.html#clustering-and-interpretation" class="external-link">ADT-based clustering and interpretation</a>
</li>
<li>review <a href="http://bioconductor.org/books/3.17/OSCA.advanced/integrating-with-protein-abundance.html#finding-correlations-between-features" class="external-link">correlations between abundances of mRNA and surface proteins</a>
</li>
<li>Examine the totalVI-based quantifications for similar findings<ul>
<li>use plotUMAP and adtProfiles</li>
<li>understand graduated relationships between surface protein and mRNA abundance</li>
</ul>
</li>
<li>Assess the sensitivity of totalVI-based interpretations to details of autoencoder training</li>
</ul>
</li>
<li>Use BiocHail and spark<ul>
<li>examine an artificial GWAS with 1000 genomes genotypes and a fabricated phenotype</li>
<li>understand how to use telomere-to-telomere variant calls with Hail</li>
</ul>
</li>
</ul>
</li>
</ul>
<!--
An example for a 45-minute workshop:

| Activity                     | Time |
|------------------------------|------|
| Packages                     | 15m  |
| Package Development          | 15m  |
| Contributing to Bioconductor | 5m   |
| Best Practices               | 10m  |

### Workshop goals and objectives

List "big picture" student-centered workshop goals and learning
objectives. Learning goals and objectives are related, but not the
same thing. These goals and objectives will help some people to decide
whether to attend the conference for training purposes, so please make
these as precise and accurate as possible.

*Learning goals* are high-level descriptions of what
participants will learn and be able to do after the workshop is
over. *Learning objectives*, on the other hand, describe in very
specific and measurable terms specific skills or knowledge
attained. The [Bloom's Taxonomy](#bloom) may be a useful framework
for defining and describing your goals and objectives, although there
are others.

### Learning goals

Some examples:

* describe how to...
* identify methods for...
* understand the difference between...

### Learning objectives

* analyze xyz data to produce...
* create xyz plots
* evaluate xyz data for artifacts

## Workshop

Divide the workshop into sections (`## A Section`). Include
fully-evaluated _R_ code chunks. Develop exercises and solutions, and
anticipate that your audience will walk through the code with you, or
work on the code idependently -- do not be too ambitious in the
material that you present.

-->
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/vjcitn/BiocPyInterop/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/vjcitn/BiocPyInterop/issues/new/choose" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing BiocPyInterop</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Vincent Carey <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-4046-0063" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Vincent Carey.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
