<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>BiocPyInterop workshop, Bioc2023 • BiocPyInterop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="BiocPyInterop workshop, Bioc2023">
<meta property="og:description" content="BiocPyInterop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BiocPyInterop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/BiocPyInterop.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/seandavi/BuildABiocWorkshop" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>BiocPyInterop workshop, Bioc2023</h1>
                        <h4 data-toc-skip class="author">Vince Carey<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/vjcitn/BiocPyInterop/blob/HEAD/vignettes/BiocPyInterop.Rmd" class="external-link"><code>vignettes/BiocPyInterop.Rmd</code></a></small>
      <div class="hidden name"><code>BiocPyInterop.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="approaches-to-reuse-of-python-based-systems-for-single-cell-genomics-and-genetics-with-bioconductor-and-basilisk">Approaches to reuse of python-based systems for single cell genomics
and genetics with Bioconductor and basilisk<a class="anchor" aria-label="anchor" href="#approaches-to-reuse-of-python-based-systems-for-single-cell-genomics-and-genetics-with-bioconductor-and-basilisk"></a>
</h2>
<p>Authors: Vince Carey<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>, <br></p>
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<div class="section level4">
<h4 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h4>
<p>Abstract: Multilingual data science strategies can increase
efficiency of discovery by taking advantage of diverse data management
and analysis strategies.</p>
<p>In this workshop we will examine interplay between R, Python, and
Apache Spark in genetic and single-cell applications. CITE-seq studies
simultaneously quantify surface protein and mRNA abundance in single
cells. We will use scviR to compare interpretations based on deep
learning and sequential component-specific methods.</p>
<p>The UK Biobank is the foundation of thousands of genome-wide
association studies. The Telomere-to-Telomere project produced the first
gapless human reference genome. Both of these resources will be explored
using BiocHail. Workshop attendees will acquire an understanding of
Aaron Lun’s <a href="https://bioconductor.org/packages/basilisk" class="external-link">basilisk</a> package
and its use in isolating specific collections of python modules, the
anndata representations and scvi-tools analyses of CITE-seq data, and
the hail.is approach to structuring and analyzing massive genetics data
resources using Spark Resilient Distributed Data. All programming will
be carried out in R; quarto documents that mix R and python will also be
illustrated.</p>
</div>
<div class="section level4">
<h4 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h4>
<ul>
<li>Basic knowledge of R syntax</li>
<li>Interest in single-cell genomics, human genetics, deep learning</li>
</ul>
<p>It will be helpful to have an acquaintance with</p>
<ul>
<li><a href="http://bioconductor.org/books/3.17/OSCA.advanced/integrating-with-protein-abundance.html" class="external-link">a
chapter of the OSCA book</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/scviR/inst/doc/citeseq_tut.html" class="external-link">an
scviR vignette</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/BiocHail/inst/doc/gwas_tut.html" class="external-link">a
BiocHail vignette</a></li>
<li>
<a href="https://vjcitn.github.io/BiocT2T/" class="external-link">a look at BiocT2T</a> –
note that we will not work with the T2T 1KG extract in detail, as it
involves a 40+ GB download, but the mechanics of working with it on your
own will be explained</li>
</ul>
</div>
<div class="section level4">
<h4 id="participation">Participation<a class="anchor" aria-label="anchor" href="#participation"></a>
</h4>
<p>This is a 90 minute workshop that will cover</p>
<ul>
<li>programming with basilisk to establish predictable python
infrastructure and interoperation</li>
<li>exploration of torch-based tooling for single-cell analysis of a
CITE-seq experiment</li>
<li>exploration of spark-based tooling for interaction with 1000 genomes
genotypes (and, if time permits, UK Biobank phenotypes)</li>
</ul>
</div>
<div class="section level4">
<h4 id="r-bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r-bioconductor-packages-used"></a>
</h4>
<ul>
<li>basilisk</li>
<li>OSCA.advanced</li>
<li>scviR</li>
<li>BiocHail</li>
</ul>
</div>
<div class="section level4">
<h4 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h4>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Verify setup</td>
<td>5m</td>
</tr>
<tr class="even">
<td>Motivations/discussion</td>
<td>5m</td>
</tr>
<tr class="odd">
<td>basilisk and basilisk.utils</td>
<td>10m</td>
</tr>
<tr class="even">
<td>(concerns with bloat)</td>
<td></td>
</tr>
<tr class="odd">
<td>OSCA advanced: CITE-seq</td>
<td>15m</td>
</tr>
<tr class="even">
<td>break</td>
<td>10 m</td>
</tr>
<tr class="odd">
<td>scviR – AnnData and tutorial VAE</td>
<td>20m</td>
</tr>
<tr class="even">
<td>Hail: exploring 1KG and UKBB</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Review</td>
<td>5m</td>
</tr>
</tbody>
</table>
<!--
An example for a 45-minute workshop:

| Activity                     | Time |
|------------------------------|------|
| Packages                     | 15m  |
| Package Development          | 15m  |
| Contributing to Bioconductor | 5m   |
| Best Practices               | 10m  |

### Workshop goals and objectives

List "big picture" student-centered workshop goals and learning
objectives. Learning goals and objectives are related, but not the
same thing. These goals and objectives will help some people to decide
whether to attend the conference for training purposes, so please make
these as precise and accurate as possible.

*Learning goals* are high-level descriptions of what
participants will learn and be able to do after the workshop is
over. *Learning objectives*, on the other hand, describe in very
specific and measurable terms specific skills or knowledge
attained. The [Bloom's Taxonomy](#bloom) may be a useful framework
for defining and describing your goals and objectives, although there
are others.

### Learning goals

Some examples:

* describe how to...
* identify methods for...
* understand the difference between...

### Learning objectives

* analyze xyz data to produce...
* create xyz plots
* evaluate xyz data for artifacts

## Workshop

Divide the workshop into sections (`## A Section`). Include
fully-evaluated _R_ code chunks. Develop exercises and solutions, and
anticipate that your audience will walk through the code with you, or
work on the code idependently -- do not be too ambitious in the
material that you present.

-->
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p><a href="mailto:stvjc@channing.harvard.edu" class="email">stvjc@channing.harvard.edu</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Harvard Medical School<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Vincent Carey.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
